File: LinkedList/Krishnan/DoublyLinkedList.cs
#https://repl.it/Et5s/2

using System;
class MainClass {
  public static void Main (string[] args) 
  {
    DoublyLinkedList<int> List = new DoublyLinkedList<int>(0);
    Console.WriteLine ("DoublyLinkedList");
    List.add(1);
    List.add(2);
    List.add(3);
    List.print();
    Console.WriteLine ("--------");
    List.delete(2);
    List.print();
    List.delete(3);
    Console.WriteLine ("--------");
    List.print();
    
  }
}

class DoublyLinkedList <T>
{
  T data;
  DoublyLinkedList<T> next;
  DoublyLinkedList<T> previous;
  
  public DoublyLinkedList(T data)
  {
    this.data = data;
  }
  
  public void add(T data)
  {
    DoublyLinkedList<T> holder = this;
    while(holder.next!= null)
    {
      holder = holder.next;
    }
    DoublyLinkedList<T> temp = new DoublyLinkedList<T>(data);
    holder.next = temp;
    temp.previous = holder;
  }
  
  public void print()
  {
    DoublyLinkedList<T> holder = this;
    while(holder.next!= null)
    {
      Console.WriteLine(holder.data);
      holder = holder.next;
    }
    Console.WriteLine(holder.data);
  }
  
  public void delete(T data)
  {
    DoublyLinkedList<T> holder = this;
    while(!holder.data.Equals(data))
    {
      holder = holder.next;
    }
    if(holder.next != null && holder.previous != null )
    {
    holder.previous.next = holder.next;
    holder.next.previous = holder.previous;
    }
    if(holder.previous == null)
    {
      holder.next.previous = holder.previous;
    }
    if(holder.next == null)
    {
      holder.previous.next = holder.next;
    }
  }
}     

File: LinkedList/Puneet/LinkedList/src/com/puneet/user/DoublyLinkedList.java
package com.puneet.user;

public class DoublyLinkedList {

	
	public static void main(String args[]){
		
		long start = System.currentTimeMillis();;
		
		DoublyLinkedList list = new DoublyLinkedList();
		list.addToFront("Puneet");
		list.addToFront("Kaku");
		list.addToFront("Milford");
		list.addToEnd("Virat");
		list.addToEnd("Sachin");
		list.addToEnd("Kamran");
		list.deleteFront();
		list.deleteEnd();
		list.traverse();
	    long end = System.currentTimeMillis();;

	    System.out.println("Execution Time : "+(end - start) + " ms");
	    
	}
	private Node head;
	
	public DoublyLinkedList(){
	    head = new Node(null);
	    head.next = head.previous = null;
	}
	  
	public Node createNode(String name){
		Node newNode = new Node(name);
		return newNode;
	}
	  
	public void addToFront(String name){
	    
	Node newNode = createNode(name);
	    
		if(head.next == null){
			head.next = newNode;
			System.out.println("New node added : "+name);
		}
		else{
	        head.next.previous = newNode;
	        newNode.next = head.next;
	        head.next = newNode;
	        System.out.println("New node added : "+name);
		}
	}
	  
	  public void addToEnd(String name){
	  
	    Node newNode = createNode(name);
	    
	    if(head.next == null){
	      head.next = newNode;
	      System.out.println("New node added : "+name);
	    }
	    
	    else{
	      Node temp = head;
	      while(temp.next != null){
	        
	        temp = temp.next;
	      }
	      temp.next = newNode;
	      System.out.println("New node added : "+name);
	    }
	  }
	  
	  public void deleteFront(){
	    
	    if(head.next == null){
	      System.out.println("The Linked List is empty");
	    }
	    else{
	      
	      Node nodeToBeDeleted = head.next;
	      System.out.println("The deleted node : "+nodeToBeDeleted.name);
	      head.next = head.next.next;
	      head.next.previous = null;
	      //delete(nodeToBeDeleted);
	    }
	  }
	  
	  public void deleteEnd(){
	    if(head.next == null){
	      System.out.println("The Linked List is empty");
	    }
	    else{
	      Node temp = head.next;
	      while(temp.next.next != null){
	        temp = temp.next;
	      }
	      Node nodeToBeDeleted = temp;
	      System.out.println("The deleted node : "+nodeToBeDeleted.next.name);
	      nodeToBeDeleted.next = null;
	    }
	  }
	  
	  public void traverse(){
		  
		  Node temp = head;
		  if(head == null || temp.next == null){
			  System.out.println("The list is already empty");
			  return;
		  }
		  
		  System.out.println("Starting Linked List Traversal");
		  temp = temp.next;
		  while(temp != null){
			  
			  System.out.println(temp.name);
			  temp = temp.next;
		  }
	  }
}
